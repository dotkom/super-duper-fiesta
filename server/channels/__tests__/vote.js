jest.mock('../../models/user');
jest.mock('../../models/vote');
jest.mock('../../models/issue');
jest.mock('../../models/meeting');
const { submitRegularVote, submitAnonymousVote } = require('../vote');
const { haveIVoted, createVote } = require('../../models/vote');
const { getIssueById } = require('../../models/issue');
const { getActiveGenfors, getGenfors } = require('../../models/meeting');
const { getAnonymousUser } = require('../../models/user');
const { generateIssue, generateSocket } = require('../../utils/generateTestData');
const { CAN_VOTE, IS_LOGGED_IN } = require('../../../common/auth/permissions');

getIssueById.mockImplementation(async () => generateIssue());
haveIVoted.mockImplementation(async () => false);
createVote.mockImplementation(() => ({
  save: async () => jest.fn(),
}));
getActiveGenfors.mockImplementation(async () => ({
  id: '1',
}));
getGenfors.mockImplementation(async id => ({
  id,
}));
getAnonymousUser.mockImplementation(async () => ({
  _id: '1',
}));

const generateData = () => ({
  issue: '1',
  alternative: '1',
});


describe('submitRegularVote', () => {
  it('emits error when not registered', async () => {
    const socket = generateSocket({ completedRegistration: false });
    await submitRegularVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });

  it('broadcasts nothing when not registered', async () => {
    const socket = generateSocket({ completedRegistration: false });
    await submitRegularVote(socket, generateData());

    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits vote', async () => {
    const socket = generateSocket({ completedRegistration: true });
    await submitRegularVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });

  it('broadcasts vote', async () => {
    const socket = generateSocket({ completedRegistration: true });
    await submitRegularVote(socket, generateData());

    expect(socket.broadcast.emit.mock.calls).toMatchSnapshot();
  });

  it('emits error with insufficient permissions', async () => {
    const socket = generateSocket({ completedRegistration: true, permissions: IS_LOGGED_IN });
    await submitRegularVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });

  it('broadcasts nothing when voting with insufficient permissions', async () => {
    const socket = generateSocket({ completedRegistration: true, permissions: IS_LOGGED_IN });
    await submitRegularVote(socket, generateData());

    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits error when trying to vote with canVote false', async () => {
    const socket = generateSocket({
      completedRegistration: true,
      permissions: CAN_VOTE,
      canVote: false,
    });

    await submitRegularVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });
});

describe('submitAnonymousVote', () => {
  it('emits error when not registered', async () => {
    const socket = generateSocket({ completedRegistration: false });
    await submitAnonymousVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });

  it('broadcasts nothing when not registered', async () => {
    const socket = generateSocket({ completedRegistration: false });
    await submitAnonymousVote(socket, generateData());

    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits error with insufficient permissions', async () => {
    const socket = generateSocket({ completedRegistration: true, permissions: IS_LOGGED_IN });
    await submitAnonymousVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });

  it('broadcasts nothing when insufficient permissions', async () => {
    const socket = generateSocket({ completedRegistration: true, permissions: IS_LOGGED_IN });
    await submitAnonymousVote(socket, generateData());

    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });


  it('emits vote', async () => {
    const socket = generateSocket({ completedRegistration: true });
    await submitAnonymousVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
  });


  it('emits error when trying to vote with canVote false', async () => {
    const socket = generateSocket({
      completedRegistration: true,
      permissions: CAN_VOTE,
      canVote: false,
    });

    await submitAnonymousVote(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });
});
