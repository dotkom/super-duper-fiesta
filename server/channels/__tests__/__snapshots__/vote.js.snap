// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submitAnonymousVote emits error when not registered 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Du er ikke registert",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitAnonymousVote emits error when trying to vote with canVote false 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Du har ikke stemmerett",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitAnonymousVote emits error with insufficient permissions 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Brukeren har ikke riktig rettigheter",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitAnonymousVote emits vote 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "_id": undefined,
        "alternative": undefined,
        "question": undefined,
        "user": undefined,
      },
      "type": "ADD_VOTE",
    },
  ],
  Array [
    "action",
    Object {
      "data": Object {
        "voted": true,
      },
      "type": "VOTED_STATE",
    },
  ],
]
`;

exports[`submitRegularVote broadcasts vote 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "_id": undefined,
        "alternative": undefined,
        "question": undefined,
        "user": undefined,
      },
      "type": "ADD_VOTE",
    },
  ],
]
`;

exports[`submitRegularVote emits error when not registered 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Du er ikke registert",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitRegularVote emits error when trying to vote with canVote false 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Du har ikke stemmerett",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitRegularVote emits error with insufficient permissions 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "id": 0,
        "message": "Brukeren har ikke riktig rettigheter",
      },
      "type": "ERROR",
    },
  ],
]
`;

exports[`submitRegularVote emits vote 1`] = `
Array [
  Array [
    "action",
    Object {
      "data": Object {
        "_id": undefined,
        "alternative": undefined,
        "question": undefined,
        "user": undefined,
      },
      "type": "ADD_VOTE",
    },
  ],
  Array [
    "action",
    Object {
      "data": Object {
        "voted": true,
      },
      "type": "VOTED_STATE",
    },
  ],
]
`;
